core/capabilities/triggers/logevent/trigger.go

```diff
@@ -0,0 +1,202 @@
+	if err != nil {
+		return nil, nil,
+			fmt.Errorf("error fetching contractReader for chainID %d from relayerSet: %v", logEventConfig.ChainID, err)
+	}
+
+	// Bind Contract in ContractReader
+	boundContracts := []types.BoundContract{{Name: reqConfig.ContractName, Address: reqConfig.ContractAddress}}
+	err = contractReader.Bind(ctx, boundContracts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Get current block HEAD/tip of the blockchain to start polling from
+	latestHead, err := relayer.LatestHead(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error getting latestHead from relayer client: %v", err)
+	}
+	height, err := strconv.ParseUint(latestHead.Height, 10, 64)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid height in latestHead from relayer client: %v", err)
+	}
+	startBlockNum := uint64(0)
+	if height > logEventConfig.LookbackBlocks {
+		startBlockNum = height - logEventConfig.LookbackBlocks
+	}
+
+	// Setup callback channel, logger and ticker to poll ContractReader
+	callbackCh := make(chan capabilities.TriggerResponse, defaultSendChannelBufferSize)
+	ticker := time.NewTicker(time.Duration(logEventConfig.PollPeriod) * time.Millisecond)
+	done := make(chan bool)
+	lggr, err := logger.New()
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not initialise logger for LogEventTrigger")
+	}
+
+	// Initialise a Log Event Trigger
+	l := &logEventTrigger{
+		ch:   callbackCh,
+		lggr: logger.Named(lggr, "LogEventTrigger: "),
+
+		reqConfig:      reqConfig,
+		contractReader: contractReader,
+		relayer:        relayer,
+		startBlockNum:  startBlockNum,
+
+		logEventConfig: logEventConfig,
+		ticker:         ticker,
+		done:           done,
+	}
+	go l.Listen(ctx)
+
+	return l, callbackCh, nil
+}
+
+// Listen to contract events and trigger workflow runs
+func (l *logEventTrigger) Listen(ctx context.Context) {
+	// Listen for events from lookbackPeriod
+	var logs []types.Sequence
+	var err error
+	logData := make(map[string]any)
+	cursor := ""
+	limitAndSort := query.LimitAndSort{
+		SortBy: []query.SortBy{query.NewSortByTimestamp(query.Asc)},
+	}
+	for {
+		select {
+		case <-l.done:
+			return
+		case t := <-l.ticker.C:
+			l.lggr.Infow("Polling event logs from ContractReader using QueryKey at", "time", t,
+				"startBlockNum", l.startBlockNum,
+				"cursor", cursor)
+			if cursor != "" {
+				limitAndSort.Limit = query.Limit{Cursor: cursor}
+			}
+			logs, err = l.contractReader.QueryKey(
+				ctx,
+				types.BoundContract{Name: l.reqConfig.ContractName, Address: l.reqConfig.ContractAddress},
+				query.KeyFilter{
+					Key: l.reqConfig.ContractEventName,
+					Expressions: []query.Expression{
+						query.Confidence(primitives.Unconfirmed),
+						query.Block(fmt.Sprintf("%d", l.startBlockNum), primitives.Gte),
+					},
+				},
+				limitAndSort,
+				&logData,
+			)
+			if err != nil {
+				l.lggr.Fatalw("QueryKey failure", "err", err)
+				continue
+			}
+			if len(logs) == 1 && logs[0].Cursor == cursor {
```

Comment:
Can we add a comment explaining what the expected response is here? It's odd to me that we get 1 log back and are logging "No new logs since...".

Reply:
Added comment